crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/generated_acir.rs

```diff
@@ -190,6 +190,151 @@ impl GeneratedAcir {
         Ok(limb_witnesses)
     }
 
+    /// Computes lhs/rhs by using euclidean division.
+    ///
+    /// Returns `q` for quotient and `r` for remainder such
+    /// that lhs = rhs * q + r
+    pub(crate) fn euclidean_division(
+        &mut self,
+        lhs: &Expression,
+        rhs: &Expression,
+        bit_size: u32,
+        predicate: &Expression,
+    ) -> Result<(Witness, Witness), AcirGenError> {
+        let q_witness = self.next_witness_index();
+        let r_witness = self.next_witness_index();
+
+        let pa = lhs * predicate;
+        self.push_opcode(AcirOpcode::Directive(Directive::Quotient(QuotientDirective {
+            a: lhs.clone(),
+            b: rhs.clone(),
+            q: q_witness,
+            r: r_witness,
+            predicate: Some(predicate.clone()),
+        })));
+
+        //r<b
+        let r_expr = Expression::from(r_witness);
+        self.range_constraint(r_witness, bit_size)?;
+        self.bound_constraint_with_offset(&r_expr, rhs, predicate, bit_size)?;
+        //range check q<=a
+        self.range_constraint(q_witness, bit_size)?;
+        // a-b*q-r = 0
+        let mut d = rhs * &Expression::from(q_witness);
+        d = &d + r_witness;
+        d = &d * predicate;
+        let div_euclidean = &pa - &d;
+
+        self.push_opcode(AcirOpcode::Arithmetic(div_euclidean));
+
+        Ok((q_witness, r_witness))
+    }
+
+    /// Generate constraints that are satisfied iff
+    /// a < b , when offset is 1, or
+    /// a <= b, when offset is 0
+    /// bits is the bit size of a and b (or an upper bound of the bit size)
+    ///
+    /// a<=b is done by constraining b-a to a bit size of 'bits':
+    /// if a<=b, 0 <= b-a <= b < 2^bits
+    /// if a>b, b-a = p+b-a > p-2^bits >= 2^bits  (if log(p) >= bits + 1)
+    /// n.b: we do NOT check here that a and b are indeed 'bits' size
+    /// a < b <=> a+1<=b
+    /// TODO: Consolidate this with bounds_check function.
```

Comment:
Open an issue for this: In particular, we want to consolidate the logic around bounds_check, truncate, modulo and unsigned division.

truncate and modulo right now use two different codepaths but they should not need to

Reply:
Issue: https://github.com/noir-lang/noir/issues/1538

Reply:
This function is a pig to read, but it can be tidied up as part of GH-1538.